fpga_interchange/fpga_interchange.cpp

```diff
@@ -0,0 +1,826 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2021  Symbiflow Authors
+ *
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "fpga_interchange.h"
+#include <capnp/message.h>
+#include <sstream>
+#include <capnp/serialize.h>
+#include <kj/std/iostream.h>
+#include "PhysicalNetlist.capnp.h"
+#include "LogicalNetlist.capnp.h"
+#include "zlib.h"
+#include "frontend_base.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+static void write_message(::capnp::MallocMessageBuilder & message, const std::string &filename) {
+    kj::Array<capnp::word> words = messageToFlatArray(message);
+    kj::ArrayPtr<kj::byte> bytes = words.asBytes();
+
+    gzFile file = gzopen(filename.c_str(), "w");
+    NPNR_ASSERT(file != Z_NULL);
+
+    NPNR_ASSERT(gzwrite(file, &bytes[0], bytes.size()) == bytes.size());
+    NPNR_ASSERT(gzclose(file) == Z_OK);
+}
+
+struct StringEnumerator {
+    std::vector<std::string> strings;
+    std::map<std::string, size_t> string_to_index;
```

Comment:
is there a reason here for using `map` and not `unordered_map`?

Reply:
No reason, changed.